atc/db/build_in_memory_check_test.go

```diff
@@ -245,6 +254,14 @@ var _ = Describe("Build", func() {
 						Expect(err).ToNot(HaveOccurred())
 					})
 
+					It("update in-memory build status for resource", func() {
+						reloaded, err := defaultResource.Reload()
+						Expect(reloaded).To(BeTrue())
+						Expect(err).ToNot(HaveOccurred())
+
+						Expect(defaultResource.BuildSummary().Status).To(Equal(atc.StatusSucceeded))
+					})
+
```

Comment:
I think we need to add two test cases when `OnCheckBuildStart` is not called:

1. Finish with "succeeded", no db operation should be done, this is for non-run checks.
2. Finish with an error, db operation should be done, this is for the problem this pr is fixing.

Reply:
I am a bit confused, when you say `non-run checks` do you mean a succeeded check build that saves its build events to memory? If yes, call it `non-db checks` might be better.

`non-run checks` to me means a check build that picked up by engine but failed to run the actual `check` script in `exec/check_step.go`, like the check with var error in this case.

Reply:
I meant the cases where `checkDelegate.WaitToRun()` returns false, thus no need to start a container and run check.

Reply:
> `checkDelegate.WaitToRun()` returns false

I suppose you meant this here
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/exec/check_step.go#L146

if `run` is fales, it goes to else block to store the found version and then
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/exec/check_step.go#L233

since check delegate doesn have an implementation of `Finished`, it goes to the one in build step delegate
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/engine/build_step_delegate.go#L151

which will then save the finish event to the build using `SaveEvent` here
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/db/build_in_memory_check.go#L518

I might be missing some important part, but the non-run checks you meant are not relevant to the `Finish` method that I changed. Could you give some more details?